WalletWasabi.Gui/Tabs/WalletManager/LoadWalletViewModel.cs

```diff
@@ -440,122 +394,109 @@ public async Task<KeyManager> LoadKeyManagerAsync(bool requirePassword, bool isH
 				var walletName = selectedWallet.WalletName;
 				if (isHardwareWallet)
 				{
-					if (selectedWallet.HardwareWalletInfo is null)
+					using (var cts = new CancellationTokenSource(TimeSpan.FromMinutes(3)))
 					{
-						SetValidationMessage("No hardware wallets detected.");
-						return null;
-					}
+						var client = new HwiClient(Global.Network);
 
-					if (!selectedWallet.HardwareWalletInfo.Initialized)
-					{
-						IEnumerable<HardwareWalletInfo> hwis;
-						try
+						if (selectedWallet.HardwareWalletInfo is null)
 						{
-							IsHardwareBusy = true;
-							MainWindowViewModel.Instance.StatusBar.TryAddStatus(StatusBarStatus.SettingUpHardwareWallet);
-							if (!await HwiProcessManager.SetupAsync(selectedWallet.HardwareWalletInfo))
+							SetValidationMessage("No hardware wallets detected.");
+							return null;
+						}
+
+						if (!selectedWallet.HardwareWalletInfo.IsInitialized())
+						{
+							try
+							{
+								IsHardwareBusy = true;
+								MainWindowViewModel.Instance.StatusBar.TryAddStatus(StatusBarStatus.SettingUpHardwareWallet);
+
+								// Setup may take a while for users to write down stuff.
+								using (var ctsSetup = new CancellationTokenSource(TimeSpan.FromMinutes(21)))
```

Comment:
Shouldn't this be a linked cancellation token?

Reply:
Linked to what?